# Exercise 2: Creating Redshift Cluster using the AWS python SDK 

## Chapter 1
### An example of Infrastructure-as-code

In [106]:
import pandas as pd
import boto3
import json
from time import time
import matplotlib.pyplot as plt
import matplotlib

### STEP 0: Make sure you have an AWS secret and access key

- Create a new IAM user in your AWS account
- Give it `AdministratorAccess`, From `Attach existing policies directly` Tab
- Take note of the access key and secret 
- Edit the file `dwh.cfg` in the same folder as this notebook and fill
<font color='red'>
<BR>
[AWS]<BR>
KEY= YOUR_AWS_KEY<BR>
SECRET= YOUR_AWS_SECRET<BR>
<font/>


In [75]:
import configparser
config = configparser.ConfigParser()
config.read_file(open('dwh.cfg'))

KEY                    = config.get('AWS','KEY')
SECRET                 = config.get('AWS','SECRET')

DWH_CLUSTER_TYPE       = config.get("DWH","DWH_CLUSTER_TYPE")
DWH_NUM_NODES          = config.get("DWH","DWH_NUM_NODES")
DWH_NODE_TYPE          = config.get("DWH","DWH_NODE_TYPE")

DWH_CLUSTER_IDENTIFIER = config.get("DWH","DWH_CLUSTER_IDENTIFIER")
DWH_DB                 = config.get("DWH","DWH_DB")
DWH_DB_USER            = config.get("DWH","DWH_DB_USER")
DWH_DB_PASSWORD        = config.get("DWH","DWH_DB_PASSWORD")
DWH_PORT               = config.get("DWH","DWH_PORT")

DWH_IAM_ROLE_NAME      = config.get("DWH", "DWH_IAM_ROLE_NAME")

(DWH_DB_USER, DWH_DB_PASSWORD, DWH_DB)

pd.DataFrame({"Param":
                  ["DWH_CLUSTER_TYPE", "DWH_NUM_NODES", "DWH_NODE_TYPE", "DWH_CLUSTER_IDENTIFIER", "DWH_DB", "DWH_DB_USER", "DWH_DB_PASSWORD", "DWH_PORT", "DWH_IAM_ROLE_NAME"],
              "Value":
                  [DWH_CLUSTER_TYPE, DWH_NUM_NODES, DWH_NODE_TYPE, DWH_CLUSTER_IDENTIFIER, DWH_DB, DWH_DB_USER, DWH_DB_PASSWORD, DWH_PORT, DWH_IAM_ROLE_NAME]
             })

Param        Value
0  DWH_CLUSTER_TYPE        multi-node 
1  DWH_NUM_NODES           4          
2  DWH_NODE_TYPE           dc2.large  
3  DWH_CLUSTER_IDENTIFIER  dwh-cluster
4  DWH_DB                  dwh        
5  DWH_DB_USER             dwhuser    
6  DWH_DB_PASSWORD         Passw0rd   
7  DWH_PORT                5439       
8  DWH_IAM_ROLE_NAME       dwh-role

### STEP 1: Creating iam client

In [76]:
# client
iam = boto3.client('iam',aws_access_key_id=KEY,
                     aws_secret_access_key=SECRET,
                     region_name='us-west-2'
                  )

### **iam.create_role**  
Creates a new role for your AWS account. For more information about roles, go to IAM Roles . For information about limitations on role names and the number of roles you can create, go to Limitations on IAM Entities in the IAM User Guide .

### Parameters
- **Path** (string) --
The path to the role. For more information about paths, see IAM Identifiers in the IAM User Guide . This parameter is optional. If it is not included, it defaults to a slash (/). This parameter allows (through its regex pattern ) a string of characters consisting of either a forward slash (/) by itself or a string that must begin and end with forward slashes. In addition, it can contain any ASCII character from the ! (u0021) through the DEL character (u007F), including most punctuation characters, digits, and upper and lowercased letters.

- **RoleName** (string) --
**[REQUIRED]**
The name of the role to create. IAM user, group, role, and policy names must be unique within the account. Names are not distinguished by case. For example, you cannot create resources named both "MyResource" and "myresource".

- **AssumeRolePolicyDocument** (string) --
**[REQUIRED]**
The trust relationship policy document that grants an entity permission to assume the role. You must provide policies in JSON format in IAM. However, for AWS CloudFormation templates formatted in YAML, you can provide the policy in JSON or YAML format. AWS CloudFormation always converts a YAML policy to JSON format before submitting it to IAM. 
The regex pattern used to validate this parameter is a string of characters consisting of the following:-
Any printable ASCII character ranging from the space character (u0020) through the end of the ASCII character range
The printable characters in the Basic Latin and Latin-1 Supplement character set (through u00FF)
The special characters tab (u0009), line feed (u000A), and carriage return (u000D)

- **Description** (string) -- A description of the role.
- **MaxSessionDuration** (integer) --
- **The maximum session duration** (in seconds) -- that you want to set for the specified role. If you do not specify a value for this setting, the default maximum of one hour is applied. This setting can have a value from 1 hour to 12 hours. Anyone who assumes the role from the AWS CLI or API can use the DurationSeconds API parameter or the duration-seconds CLI parameter to request a longer session. The MaxSessionDuration setting determines the maximum duration that can be requested using the DurationSeconds parameter. If users don't specify a value for the DurationSeconds parameter, their security credentials are valid for one hour by default. This applies when you use the AssumeRole* API operations or the assume-role* CLI operations but does not apply when you use those operations to create a console URL. For more information, see Using IAM Roles in the IAM User Guide .

- **PermissionsBoundary** (string) -- The ARN of the policy that is used to set the permissions boundary for the role.
- **Tags**(list) --
A list of tags that you want to attach to the newly created role. Each tag consists of a key name and an associated value. For more information about tagging, see Tagging IAM Identities in the IAM User Guide .

---
### **iam.attach_policy**
Attaches the specified managed policy to the specified IAM group. You use this API to attach a managed policy to a group. To embed an inline policy in a group, use PutGroupPolicy.

### Parameters
**PolicyArn** (string) --
**[REQUIRED]**
The Amazon Resource Name (ARN) of the IAM policy you want to attach.

---
### **iam.get_role**
Retrieves information about the specified role, including the role's path, GUID, ARN, and the role's trust policy that grants permission to assume the role. For more information about roles, see Working with Roles.

### Parameters
**RoleName** (string) --
**[REQUIRED]**
The name of the IAM role to get information about.
This parameter allows (through its regex pattern ) a string of characters consisting of upper and lowercase alphanumeric characters with no spaces. You can also include any of the following characters: _+=,.@-

In [77]:
from botocore.exceptions import ClientError

#1.1 Create the role, 
try:
    print("1.1 Creating a new IAM Role") 
    dwhRole = iam.create_role(
        Path = '/',
        RoleName = DWH_IAM_ROLE_NAME,
        Description = "Allows Redshift clusters to call AWS services on your behalf.",
        AssumeRolePolicyDocument = json.dumps(
            {'Statement': [{'Action': 'sts:AssumeRole',
               'Effect': 'Allow',
               'Principal': {'Service': 'redshift.amazonaws.com'}}],
             'Version': '2012-10-17'})
    )    
except Exception as e:
    print(e)
    
    
print("1.2 Attaching Policy")

iam.attach_role_policy(RoleName=DWH_IAM_ROLE_NAME,
                       PolicyArn="arn:aws:iam::aws:policy/AmazonS3ReadOnlyAccess"
                      )['ResponseMetadata']['HTTPStatusCode']

print("1.3 Get the IAM role ARN")
roleArn = iam.get_role(RoleName=DWH_IAM_ROLE_NAME)['Role']['Arn']

print(roleArn)

1.1 Creating a new IAM Role
1.2 Attaching Policy
1.3 Get the IAM role ARN
arn:aws:iam::764499268961:role/dwh-role


### STEP 2:  Redshift Cluster

- Create a RedShift Cluster
- For complete arguments to `create_cluster`, see [docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/redshift.html#Redshift.Client.create_cluster)

#### Creating redshift client

In [78]:
# client
redshift = boto3.client('redshift',
                       region_name="us-west-2",
                       aws_access_key_id=KEY,
                       aws_secret_access_key=SECRET
                       )

### **redshift.create_cluster**  
Creates a new cluster. To create a cluster in Virtual Private Cloud (VPC), you must provide a cluster subnet group name. The cluster subnet group identifies the subnets of your VPC that Amazon Redshift uses when creating the cluster. For more information about managing clusters, go to Amazon Redshift Clusters in the Amazon Redshift Cluster Management Guide .


### Parameters
**DBName** (string) --
The name of the first database to be created when the cluster is created.
To create additional databases after the cluster is created, connect to the cluster with a SQL client and use SQL commands to create a database. For more information, go to Create a Database in the Amazon Redshift Database Developer Guide.
Default: dev
*Constraints*:
Must contain 1 to 64 alphanumeric characters.
Must contain only lowercase letters.
Cannot be a word that is reserved by the service. A list of reserved words can be found in Reserved Words in the Amazon Redshift Database Developer Guide.

**ClusterIdentifier** (string) --
**[REQUIRED]**
A unique identifier for the cluster. You use this identifier to refer to the cluster for any subsequent cluster operations such as deleting or modifying. The identifier also appears in the Amazon Redshift console.
*Constraints*:
Must contain from 1 to 63 alphanumeric characters or hyphens.
Alphabetic characters must be lowercase.
First character must be a letter.
Cannot end with a hyphen or contain two consecutive hyphens.
Must be unique for all clusters within an AWS account.
Example: myexamplecluster

**ClusterType** (string) --
The type of the cluster. When cluster type is specified as
single-node , the NumberOfNodes parameter is not required.
multi-node , the NumberOfNodes parameter is required.
Valid Values: multi-node | single-node
Default: multi-node
NodeType (string) --
**[REQUIRED]**
The node type to be provisioned for the cluster. For information about node types, go to Working with Clusters in the Amazon Redshift Cluster Management Guide .
Valid Values: ds2.xlarge | ds2.8xlarge | ds2.xlarge | ds2.8xlarge | dc1.large | dc1.8xlarge | dc2.large | dc2.8xlarge

**MasterUsername** (string) --
**[REQUIRED]**
The user name associated with the master user account for the cluster that is being created.
Constraints:
Must be 1 - 128 alphanumeric characters. The user name can't be PUBLIC .
First character must be a letter.
Cannot be a reserved word. A list of reserved words can be found in Reserved Words in the Amazon Redshift Database Developer Guide.

**MasterUserPassword** (string) --
**[REQUIRED]**
The password associated with the master user account for the cluster that is being created.
Constraints:
Must be between 8 and 64 characters in length.
Must contain at least one uppercase letter.
Must contain at least one lowercase letter.
Must contain one number.
Can be any printable ASCII character (ASCII code 33 to 126) except ' (single quote), " (double quote), , /, @, or space.

**ClusterSecurityGroups** (list) --
A list of security groups to be associated with this cluster.
Default: The default cluster security group for Amazon Redshift.
(string) --

**VpcSecurityGroupIds** (list) --
A list of Virtual Private Cloud (VPC) security groups to be associated with the cluster.
Default: The default VPC security group is associated with the cluster.
(string) --

**ClusterSubnetGroupName** (string) --
The name of a cluster subnet group to be associated with this cluster.
If this parameter is not provided the resulting cluster will be deployed outside virtual private cloud (VPC).

**AvailabilityZone** (string) --
The EC2 Availability Zone (AZ) in which you want Amazon Redshift to provision the cluster. For example, if you have several EC2 instances running in a specific Availability Zone, then you might want the cluster to be provisioned in the same zone in order to decrease network latency.
Default: A random, system-chosen Availability Zone in the region that is specified by the endpoint.
Example: us-east-1d
Constraint: The specified Availability Zone must be in the same region as the current endpoint.

**PreferredMaintenanceWindow** (string) --
The weekly time range (in UTC) during which automated cluster maintenance can occur.
Format: ddd:hh24:mi-ddd:hh24:mi
Default: A 30-minute window selected at random from an 8-hour block of time per region, occurring on a random day of the week. For more information about the time blocks for each region, see Maintenance Windows in Amazon Redshift Cluster Management Guide.
Valid Days: Mon | Tue | Wed | Thu | Fri | Sat | Sun
Constraints: Minimum 30-minute window.

**ClusterParameterGroupName** (string) --
The name of the parameter group to be associated with this cluster.
Default: The default Amazon Redshift cluster parameter group. For information about the default parameter group, go to Working with Amazon Redshift Parameter Groups
Constraints:
Must be 1 to 255 alphanumeric characters or hyphens.
First character must be a letter.
Cannot end with a hyphen or contain two consecutive hyphens.
AutomatedSnapshotRetentionPeriod (integer) --
The number of days that automated snapshots are retained. If the value is 0, automated snapshots are disabled. Even if automated snapshots are disabled, you can still create manual snapshots when you want with CreateClusterSnapshot.
Default: 1
Constraints: Must be a value from 0 to 35.

**ManualSnapshotRetentionPeriod** (integer) --
The default number of days to retain a manual snapshot. If the value is -1, the snapshot is retained indefinitely. This setting doesn't change the retention period of existing snapshots.
The value must be either -1 or an integer between 1 and 3,653.

**Port (integer)** --
The port number on which the cluster accepts incoming connections.
The cluster is accessible only via the JDBC and ODBC connection strings. Part of the connection string requires the port on which the cluster will listen for incoming connections.
Default: 5439
Valid Values: 1150-65535

**ClusterVersion (string)** --
The version of the Amazon Redshift engine software that you want to deploy on the cluster.
The version selected runs on all the nodes in the cluster.
Constraints: Only version 1.0 is currently available.
Example: 1.0

**AllowVersionUpgrade** (boolean) --
If true , major version upgrades can be applied during the maintenance window to the Amazon Redshift engine that is running on the cluster.
When a new major version of the Amazon Redshift engine is released, you can request that the service automatically apply upgrades during the maintenance window to the Amazon Redshift engine that is running on your cluster.
Default: true

**NumberOfNodes** (integer) --
The number of compute nodes in the cluster. This parameter is required when the ClusterType parameter is specified as multi-node .
For information about determining how many nodes you need, go to Working with Clusters in the Amazon Redshift Cluster Management Guide .
If you don't specify this parameter, you get a single-node cluster. When requesting a multi-node cluster, you must specify the number of nodes that you want in the cluster.
Default: 1
Constraints: Value must be at least 1 and no more than 100.

**PubliclyAccessible** (boolean) -- If true , the cluster can be accessed from a public network.
Encrypted (boolean) --
If true , the data in the cluster is encrypted at rest.
Default: false

**HsmClientCertificateIdentifier** (string) -- Specifies the name of the HSM client certificate the Amazon Redshift cluster uses to retrieve the data encryption keys stored in an HSM.

**HsmConfigurationIdentifier** (string) -- Specifies the name of the HSM configuration that contains the information the Amazon Redshift cluster can use to retrieve and store keys in an HSM.
ElasticIp (string) --
The Elastic IP (EIP) address for the cluster.
Constraints: The cluster must be provisioned in EC2-VPC and publicly-accessible through an Internet gateway. For more information about provisioning clusters in EC2-VPC, go to Supported Platforms to Launch Your Cluster in the Amazon Redshift Cluster Management Guide.

**Tags** (list) --
A list of tag instances.
(dict) --
A tag consisting of a name/value pair for a resource.
Key (string) --
The key, or name, for the resource tag.
Value (string) --
The value for the resource tag.

**KmsKeyId** (string) -- The AWS Key Management Service (KMS) key ID of the encryption key that you want to use to encrypt data in the cluster.
EnhancedVpcRouting (boolean) --
An option that specifies whether to create the cluster with enhanced VPC routing enabled. To create a cluster that uses enhanced VPC routing, the cluster must be in a VPC. For more information, see Enhanced VPC Routing in the Amazon Redshift Cluster Management Guide.
If this option is true , enhanced VPC routing is enabled.
Default: false

**AdditionalInfo** (string) -- Reserved.

**IamRoles** (list) --
A list of AWS Identity and Access Management (IAM) roles that can be used by the cluster to access other AWS services. You must supply the IAM roles in their Amazon Resource Name (ARN) format. You can supply up to 10 IAM roles in a single request.
A cluster can have up to 10 IAM roles associated with it at any time.
(string) --

**MaintenanceTrackName** (string) -- An optional parameter for the name of the maintenance track for the cluster. If you don't provide a maintenance track name, the cluster is assigned to the current track.

**SnapshotScheduleIdentifier** (string) -- A unique identifier for the snapshot schedule.

In [79]:
try:
    response = redshift.create_cluster(        
        #HW
        ClusterType=DWH_CLUSTER_TYPE,
        NodeType=DWH_NODE_TYPE,
        NumberOfNodes=int(DWH_NUM_NODES),

        #Identifiers & Credentials
        DBName=DWH_DB,
        ClusterIdentifier=DWH_CLUSTER_IDENTIFIER,
        MasterUsername=DWH_DB_USER,
        MasterUserPassword=DWH_DB_PASSWORD,
        
        #Roles (for s3 access)
        IamRoles=[roleArn]  
    )
except Exception as e:
    print(e)

In [80]:
def prettyRedshiftProps(props):
    pd.set_option('display.max_colwidth', -1)
    keysToShow = ["ClusterIdentifier", "NodeType", "ClusterStatus", "MasterUsername", "DBName", "Endpoint", "NumberOfNodes", 'VpcId']
    x = [(k, v) for k,v in props.items() if k in keysToShow]
    return pd.DataFrame(data=x, columns=["Key", "Value"])

### **redshift.describe_clusters**  
Returns properties of provisioned clusters including general cluster properties, cluster database properties, maintenance and backup properties, and security and access properties. This operation supports pagination. For more information about managing clusters, go to Amazon Redshift Clusters in the Amazon Redshift Cluster Management Guide .

If you specify both tag keys and tag values in the same request, Amazon Redshift returns all clusters that match any combination of the specified keys and values. For example, if you have owner and environment for tag keys, and admin and test for tag values, all clusters that have any combination of those values are returned.

If both tag keys and values are omitted from the request, clusters are returned regardless of whether they have tag keys or values associated with them.

### Parameters
**ClusterIdentifier** (string) --
The unique identifier of a cluster whose properties you are requesting. This parameter is case sensitive.

The default is that all clusters defined for an account are returned.

**MaxRecords** (integer) --
The maximum number of response records to return in each call. If the number of remaining response records exceeds the specified MaxRecords value, a value is returned in a marker field of the response. You can retrieve the next set of records by retrying the command with the returned marker value.
Default: 100
Constraints: minimum 20, maximum 100.

**Marker** (string) --
An optional parameter that specifies the starting point to return a set of response records. When the results of a DescribeClusters request exceed the value specified in MaxRecords , AWS returns a value in the Marker field of the response. You can retrieve the next set of response records by providing the returned marker value in the Marker parameter and retrying the request.
Constraints: You can specify either the ClusterIdentifier parameter or the Marker parameter, but not both.

**TagKeys** (list) --
A tag key or keys for which you want to return all matching clusters that are associated with the specified key or keys. For example, suppose that you have clusters that are tagged with keys called owner and environment . If you specify both of these tag keys in the request, Amazon Redshift returns a response with the clusters that have either or both of these tag keys associated with them.
(string) --

**TagValues** (list) --
A tag value or values for which you want to return all matching clusters that are associated with the specified tag value or values. For example, suppose that you have clusters that are tagged with values called admin and test . If you specify both of these tag values in the request, Amazon Redshift returns a response with the clusters that have either or both of these tag values associated with them.

In [83]:
myClusterProps = redshift.describe_clusters(ClusterIdentifier=DWH_CLUSTER_IDENTIFIER)['Clusters'][0]
prettyRedshiftProps(myClusterProps)

Key  \
0  ClusterIdentifier   
1  NodeType            
2  ClusterStatus       
3  MasterUsername      
4  DBName              
5  Endpoint            
6  VpcId               
7  NumberOfNodes       

                                                                                    Value  
0  dwh-cluster                                                                             
1  dc2.large                                                                               
2  available                                                                               
3  dwhuser                                                                                 
4  dwh                                                                                     
5  {'Address': 'dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com', 'Port': 5439}  
6  vpc-ce44eeb6                                                                            
7  4

In [19]:
myClusterProps

{'ClusterIdentifier': 'dwh-cluster',
 'NodeType': 'dc2.large',
 'ClusterStatus': 'available',
 'MasterUsername': 'dwhuser',
 'DBName': 'dwh',
 'Endpoint': {'Address': 'dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com',
  'Port': 5439},
 'ClusterCreateTime': datetime.datetime(2019, 6, 1, 12, 38, 30, 77000, tzinfo=tzutc()),
 'AutomatedSnapshotRetentionPeriod': 1,
 'ManualSnapshotRetentionPeriod': -1,
 'ClusterSecurityGroups': [],
 'VpcSecurityGroups': [],
 'ClusterParameterGroups': [{'ParameterGroupName': 'default.redshift-1.0',
   'ParameterApplyStatus': 'in-sync'}],
 'ClusterSubnetGroupName': 'default',
 'VpcId': 'vpc-ce44eeb6',
 'AvailabilityZone': 'us-west-2c',
 'PreferredMaintenanceWindow': 'fri:07:00-fri:07:30',
 'PendingModifiedValues': {},
 'ClusterVersion': '1.0',
 'AllowVersionUpgrade': True,
 'NumberOfNodes': 4,
 'PubliclyAccessible': True,
 'Encrypted': False,
 'ClusterPublicKey': 'ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCtyUAUoOAlGcYYEa98AVObSefDSLQK2/fdVTuuvKyXXpR2QD

#### Take note of the cluster endpoint and role ARN  
<font color='red'>DO NOT RUN THIS unless the cluster status becomes "Available" </font>

In [84]:
# endpoint
DWH_ENDPOINT = myClusterProps['Endpoint']['Address']
DWH_ROLE_ARN = myClusterProps['IamRoles'][0]['IamRoleArn']
print("DWH_ENDPOINT :: ", endpoint)
print("DWH_ROLE_ARN :: ", roleArn)

DWH_ENDPOINT ::  dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com
DWH_ROLE_ARN ::  arn:aws:iam::764499268961:role/dwh-role


### STEP 3: Create an endpoint to access the cluster

In [53]:
# client
ec2 = boto3.resource('ec2',
                       region_name="us-west-2",
                       aws_access_key_id=KEY,
                       aws_secret_access_key=SECRET
                    )

### class EC2.Vpc(id)
A resource representing an Amazon Elastic Compute Cloud (EC2) Vpc

---
### **security_groups**  

A collection of SecurityGroup resources  

---
### **all()**  

Creates an iterable of all SecurityGroup resources in the collection.  

---
### **authorize_ingress**  
Adds the specified ingress rules to a security group.
An inbound rule permits instances to receive traffic from the specified IPv4 or IPv6 CIDR address ranges, or from the instances associated with the specified destination security groups.
You specify a protocol for each rule (for example, TCP). For TCP and UDP, you must also specify the destination port or port range. For ICMP/ICMPv6, you must also specify the ICMP/ICMPv6 type and code. You can use -1 to mean all types or all codes.
Rule changes are propagated to instances within the security group as quickly as possible. However, a small delay might occur.

### Parameters
**CidrIp** (string) --
The IPv4 address range, in CIDR format. You can't specify this parameter when specifying a source security group. To specify an IPv6 address range, use a set of IP permissions.
Alternatively, use a set of IP permissions to specify multiple rules and a description for the rule.

**FromPort** (integer) --
The start of port range for the TCP and UDP protocols, or an ICMP type number. For the ICMP type number, use -1 to specify all types. If you specify all ICMP types, you must specify all codes.
Alternatively, use a set of IP permissions to specify multiple rules and a description for the rule.

**GroupName** (string) -- [EC2-Classic, default VPC] The name of the security group. You must specify either the security group ID or the security group name in the request.
IpPermissions (list) --
The sets of IP permissions.
(dict) --
Describes a set of permissions for a security group rule.

**FromPort** (integer) --
The start of port range for the TCP and UDP protocols, or an ICMP/ICMPv6 type number. A value of -1 indicates all ICMP/ICMPv6 types. If you specify all ICMP/ICMPv6 types, you must specify all codes.

**IpProtocol** (string) --
The IP protocol name (tcp , udp , icmp , icmpv6 ) or number (see Protocol Numbers ).

[VPC only] Use -1 to specify all protocols. When authorizing security group rules, specifying -1 or a protocol number other than tcp , udp , icmp , or icmpv6 allows traffic on all ports, regardless of any port range you specify. For tcp , udp , and icmp , you must specify a port range. For icmpv6 , the port range is optional; if you omit the port range, traffic for all types and codes is allowed.

**IpRanges** (list) --
The IPv4 ranges.
(dict) --
Describes an IPv4 range.

**CidrIp** (string) --
The IPv4 CIDR range. You can either specify a CIDR range or a source security group, not both. To specify a single IPv4 address, use the /32 prefix length.

**Description** (string) --
A description for the security group rule that references this IPv4 address range.
Constraints: Up to 255 characters in length. Allowed characters are a-z, A-Z, 0-9, spaces, and ._-:/()#,@[]+=;{}!$*

**Ipv6Ranges** (list) --
[VPC only] The IPv6 ranges.
(dict) --
[EC2-VPC only] Describes an IPv6 range.

**CidrIpv6** (string) --
The IPv6 CIDR range. You can either specify a CIDR range or a source security group, not both. To specify a single IPv6 address, use the /128 prefix length.

**Description** (string) --
A description for the security group rule that references this IPv6 address range.
Constraints: Up to 255 characters in length. Allowed characters are a-z, A-Z, 0-9, spaces, and ._-:/()#,@[]+=;{}!$*

**PrefixListIds** (list) --
[VPC only] The prefix list IDs for an AWS service. With outbound rules, this is the AWS service to access through a VPC endpoint from instances associated with the security group.
(dict) --
Describes a prefix list ID.

**Description** (string) --
A description for the security group rule that references this prefix list ID.
Constraints: Up to 255 characters in length. Allowed characters are a-z, A-Z, 0-9, spaces, and ._-:/()#,@[]+=;{}!$*

**PrefixListId** (string) --
The ID of the prefix.

**ToPort** (integer) --
The end of port range for the TCP and UDP protocols, or an ICMP/ICMPv6 code. A value of -1 indicates all ICMP/ICMPv6 codes. If you specify all ICMP/ICMPv6 types, you must specify all codes.

**UserIdGroupPairs** (list) --
The security group and AWS account ID pairs.
(dict) --
Describes a security group and AWS account ID pair.

**Description** (string) --
A description for the security group rule that references this user ID group pair.
Constraints: Up to 255 characters in length. Allowed characters are a-z, A-Z, 0-9, spaces, and ._-:/()#,@[]+=;{}!$*

**GroupId**(string) --
The ID of the security group.

**GroupName** (string) --
The name of the security group. In a request, use this parameter for a security group in EC2-Classic or a default VPC only. For a security group in a nondefault VPC, use the security group ID.
For a referenced security group in another VPC, this value is not returned if the referenced security group is deleted.

**PeeringStatus** (string) --
The status of a VPC peering connection, if applicable.

**UserId** (string) --
The ID of an AWS account.
For a referenced security group in another VPC, the account ID of the referenced security group is returned in the response. If the referenced security group is deleted, this value is not returned.
[EC2-Classic] Required when adding or removing rules that reference a security group in another AWS account.

**VpcId** (string) --
The ID of the VPC for the referenced security group, if applicable.

**VpcPeeringConnectionId** (string) --
The ID of the VPC peering connection, if applicable.

**IpProtocol** (string) --
The IP protocol name (tcp , udp , icmp ) or number (see Protocol Numbers ). To specify icmpv6 , use a set of IP permissions.
[VPC only] Use -1 to specify all protocols. If you specify -1 or a protocol other than tcp , udp , or icmp , traffic on all ports is allowed, regardless of any ports you specify.
Alternatively, use a set of IP permissions to specify multiple rules and a description for the rule.

**SourceSecurityGroupName** (string) -- [EC2-Classic, default VPC] The name of the source security group. You can't specify this parameter in combination with the following parameters: the CIDR IP address range, the start of the port range, the IP protocol, and the end of the port range. Creates rules that grant full ICMP, UDP, and TCP access. To create a rule with a specific IP protocol and port range, use a set of IP permissions instead. For EC2-VPC, the source security group must be in the same VPC.

**SourceSecurityGroupOwnerId** (string) -- [nondefault VPC] The AWS account ID for the source security group, if the source security group is in a different account. You can't specify this parameter in combination with the following parameters: the CIDR IP address range, the IP protocol, the start of the port range, and the end of the port range. Creates rules that grant full ICMP, UDP, and TCP access. To create a rule with a specific IP protocol and port range, use a set of IP permissions instead.

**ToPort** (integer) --
The end of port range for the TCP and UDP protocols, or an ICMP code number. For the ICMP code number, use -1 to specify all codes. If you specify all ICMP types, you must specify all codes.
Alternatively, use a set of IP permissions to specify multiple rules and a description for the rule.

**DryRun** (boolean) -- Checks whether you have the required permissions for the action, without actually making the request, and provides an error response. If you have the required permissions, the error response is DryRunOperation . Otherwise, it is UnauthorizedOperation 

In [29]:
try:
    vpc = ec2.Vpc(id=myClusterProps['VpcId'])
    defaultSg = list(vpc.security_groups.all())[0]
    #print(defaultSg)
    defaultSg.authorize_ingress(
        GroupName=defaultSg.group_name,
        CidrIp='0.0.0.0/0',
        IpProtocol='TCP',
        FromPort=int(DWH_PORT),
        ToPort=int(DWH_PORT)
   )
except Exception as e:
    print(e)

ec2.SecurityGroup(id='sg-d499ea99')
An error occurred (InvalidPermission.Duplicate) when calling the AuthorizeSecurityGroupIngress operation: the specified rule "peer: 0.0.0.0/0, TCP, from port: 5439, to port: 5439, ALLOW" already exists


use sql client

In [85]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [87]:
conn_string="postgresql://{}:{}@{}:{}/{}".format(DWH_DB_USER, DWH_DB_PASSWORD, DWH_ENDPOINT, DWH_PORT, DWH_DB)
print(conn_string)
%sql $conn_string

postgresql://dwhuser:Passw0rd@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh


'Connected: dwhuser@dwh'

### STEP 4: Create S3 client

In [88]:
s3 = boto3.resource('s3',
                       region_name="us-west-2",
                       aws_access_key_id=KEY,
                       aws_secret_access_key=SECRET
                    )

In [58]:
# load udacity bucket
udacity_labs_bucket = s3.Bucket('udacity-labs')
for obj in udacity_labs_bucket.objects.filter(Prefix='tickets'):
    print(obj)

s3.ObjectSummary(bucket_name='udacity-labs', key='tickets/')
s3.ObjectSummary(bucket_name='udacity-labs', key='tickets/full/')
s3.ObjectSummary(bucket_name='udacity-labs', key='tickets/full/full.csv.gz')
s3.ObjectSummary(bucket_name='udacity-labs', key='tickets/split/')
s3.ObjectSummary(bucket_name='udacity-labs', key='tickets/split/part-00000-d33afb94-b8af-407d-abd5-59c0ee8f5ee8-c000.csv.gz')
s3.ObjectSummary(bucket_name='udacity-labs', key='tickets/split/part-00001-d33afb94-b8af-407d-abd5-59c0ee8f5ee8-c000.csv.gz')
s3.ObjectSummary(bucket_name='udacity-labs', key='tickets/split/part-00002-d33afb94-b8af-407d-abd5-59c0ee8f5ee8-c000.csv.gz')
s3.ObjectSummary(bucket_name='udacity-labs', key='tickets/split/part-00003-d33afb94-b8af-407d-abd5-59c0ee8f5ee8-c000.csv.gz')
s3.ObjectSummary(bucket_name='udacity-labs', key='tickets/split/part-00004-d33afb94-b8af-407d-abd5-59c0ee8f5ee8-c000.csv.gz')
s3.ObjectSummary(bucket_name='udacity-labs', key='tickets/split/part-00005-d33afb94-b8af-407d-abd5-

#### Redshift Data Types
- SMALLINT (INT2)
- INTEGER (INT, INT4)
- BIGINT (INT8)
- DECIMAL (NUMERIC)
- REAL (FLOAT4)
- DOUBLE PRECISION (FLOAT8)
- BOOLEAN (BOOL)
- CHAR (CHARACTER)
- VARCHAR (CHARACTER VARYING)
- DATE
- TIMESTAMP
- TIMESTAMPTZ
- SERIAL (DEFAULT nextval('name_of_the_variable') NOT NULL)

#### Creating Tables

In [59]:
%%sql 
DROP TABLE IF EXISTS "sporting_event_ticket";
CREATE TABLE "sporting_event_ticket" (
    "id" double precision DEFAULT nextval('sporting_event_ticket_seq') NOT NULL,
    "sporting_event_id" double precision NOT NULL,
    "sport_location_id" double precision NOT NULL,
    "seat_level" numeric(1,0) NOT NULL,
    "seat_section" character varying(15) NOT NULL,
    "seat_row" character varying(10) NOT NULL,
    "seat" character varying(10) NOT NULL,
    "ticketholder_id" double precision,
    "ticket_price" numeric(8,2) NOT NULL
);

 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.


[]

#### Load Partitioned data into the cluster
Use the COPY command to load data from `s3://udacity-labs/tickets/split/part` using your iam role credentials. Use gzip delimiter `;`.

In [61]:
DWH_ROLE_ARN

'arn:aws:iam::764499268961:role/dwh-role'

In [63]:
%%time
qry = """
      copy sporting_event_ticket 
      from 's3://udacity-labs/tickets/split/part'
      credentials 'aws_iam_role={}'
      gzip DELIMITER ';'
      compupdate off
      region 'us-west-2';
      """.format(DWH_ROLE_ARN)

%sql $qry

 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
CPU times: user 4.53 ms, sys: 3.9 ms, total: 8.42 ms
Wall time: 28 s


In [65]:
%%sql 
DROP TABLE IF EXISTS "sporting_event_ticket_full";
CREATE TABLE "sporting_event_ticket_full" (
    "id" double precision DEFAULT nextval('sporting_event_ticket_seq') NOT NULL,
    "sporting_event_id" double precision NOT NULL,
    "sport_location_id" double precision NOT NULL,
    "seat_level" numeric(1,0) NOT NULL,
    "seat_section" character varying(15) NOT NULL,
    "seat_row" character varying(10) NOT NULL,
    "seat" character varying(10) NOT NULL,
    "ticketholder_id" double precision,
    "ticket_price" numeric(8,2) NOT NULL
);

 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.


[]

#### Inserting data full


In [66]:
%%time
qry = """
      copy sporting_event_ticket_full
      from 's3://udacity-labs/tickets/full/full.csv.gz'
      credentials 'aws_iam_role={}'
      gzip DELIMITER ';'
      compupdate off
      region 'us-west-2';
      """.format(DWH_ROLE_ARN)

%sql $qry

 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
CPU times: user 4.33 ms, sys: 2.52 ms, total: 6.85 ms
Wall time: 24 s


## Chapter 2

### Now let's work with different kinds of distribution strategies and compare them

- We are going to use a benchmarking data set common for benchmarking star schemas in data warehouses.
- The data is pre-loaded in a public bucket on the `us-west-2` region
- Our examples will be based on the Amazon Redshfit tutorial but in a scripted environment in our workspace.

<img src="./img/aws-schema.png" width="50%"/>

#### No distribution schema

In [99]:
%%sql 
DROP SCHEMA IF EXISTS nodist CASCADE;
CREATE SCHEMA nodist;
SET search_path TO nodist;

DROP TABLE IF EXISTS part cascade;
DROP TABLE IF EXISTS supplier;
DROP TABLE IF EXISTS supplier;
DROP TABLE IF EXISTS customer;
DROP TABLE IF EXISTS dwdate;
DROP TABLE IF EXISTS lineorder;

CREATE TABLE part 
(
  p_partkey       INTEGER     NOT NULL,
  p_name          VARCHAR(22) NOT NULL,
  p_mfgr          VARCHAR(6)  NOT NULL,
  p_category      VARCHAR(7)  NOT NULL,
  p_brand1        VARCHAR(9)  NOT NULL,
  p_color         VARCHAR(11) NOT NULL,
  p_type          VARCHAR(25) NOT NULL,
  p_size          INTEGER     NOT NULL,
  p_container     VARCHAR(10) NOT NULL
);

CREATE TABLE supplier 
(
    s_suppkey     INTEGER     NOT NULL,
    s_name        VARCHAR(50) NOT NULL,
    s_address     VARCHAR(50) NOT NULL,
    s_city        VARCHAR(30) NOT NULL,
    s_nation      VARCHAR(30) NOT NULL,
    s_region      VARCHAR(30) NOT NULL,
    s_phone       VARCHAR(20) NOT NULL
);

CREATE TABLE customer 
(
    c_custkey     INTEGER     NOT NULL,
    c_name        VARCHAR(50) NOT NULL,
    c_address     VARCHAR(30) NOT NULL,
    c_city        VARCHAR(30) NOT NULL,
    c_nation      VARCHAR(30) NOT NULL,
    c_region      VARCHAR(30) NOT NULL,
    c_phone       VARCHAR(30) NOT NULL,
    c_mktsegment  VARCHAR(30) NOT NULL
);

CREATE TABLE dwdate 
(
    d_datekey          INTEGER     NOT NULL,
    d_date             VARCHAR(20) NOT NULL,
    d_dayofweek        VARCHAR(20) NOT NULL,
    d_month            VARCHAR(20) NOT NULL,
    d_year             INTEGER     NOT NULL,
    d_yearmonthnum     INTEGER     NOT NULL,
    d_yearmonth        VARCHAR(10) NOT NULL,
    d_daynuminweek     INTEGER     NOT NULL,
    d_daynuminmonth    INTEGER     NOT NULL,
    d_daynuminyear     INTEGER     NOT NULL,
    d_monthnuminyear   INTEGER     NOT NULL,
    d_weeknuminyear    INTEGER     NOT NULL,
    d_sellingseason    VARCHAR(20) NOT NULL,
    d_lastdayinweekfl  VARCHAR(1)  NOT NULL,
    d_lastdayinmonthfl VARCHAR(1)  NOT NULL,
    d_holidayfl        VARCHAR(1)  NOT NULL,
    d_weekdayfl        VARCHAR(1)  NOT NULL
);
CREATE TABLE lineorder 
(
    lo_orderkey        INTEGER     NOT NULL,
    lo_linenumber      INTEGER     NOT NULL,
    lo_custkey         INTEGER     NOT NULL,
    lo_partkey         INTEGER     NOT NULL,
    lo_suppkey         INTEGER     NOT NULL,
    lo_orderdate       VARCHAR(20) NOT NULL,
    lo_orderpriority   VARCHAR(20) NOT NULL,
    lo_shippriority    VARCHAR(20) NOT NULL,
    lo_quantity        INTEGER     NOT NULL,
    lo_extendedprice   INTEGER     NOT NULL,
    lo_ordertotalprice INTEGER     NOT NULL,
    lo_discount        INTEGER     NOT NULL,
    lo_revenue         INTEGER     NOT NULL,
    lo_supplycost      INTEGER     NOT NULL,
    lo_tax             INTEGER     NOT NULL,
    lo_commitdate        VARCHAR(20) NOT NULL,
    lo_shipmode        VARCHAR(20) NOT NULL
);

 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

#### Dist schema

In [100]:
%%sql 
DROP SCHEMA IF EXISTS dist CASCADE;
CREATE SCHEMA dist;
SET search_path TO dist;

DROP TABLE IF EXISTS part cascade;
DROP TABLE IF EXISTS supplier;
DROP TABLE IF EXISTS supplier;
DROP TABLE IF EXISTS customer;
DROP TABLE IF EXISTS dwdate;
DROP TABLE IF EXISTS lineorder;

CREATE TABLE part 
(
  p_partkey       INTEGER     NOT NULL    sortkey distkey,
  p_name          VARCHAR(22) NOT NULL,
  p_mfgr          VARCHAR(6)  NOT NULL,
  p_category      VARCHAR(7)  NOT NULL,
  p_brand1        VARCHAR(9)  NOT NULL,
  p_color         VARCHAR(11) NOT NULL,
  p_type          VARCHAR(25) NOT NULL,
  p_size          INTEGER     NOT NULL,
  p_container     VARCHAR(10) NOT NULL
);

CREATE TABLE supplier 
(
    s_suppkey     INTEGER     NOT NULL    sortkey,
    s_name        VARCHAR(50) NOT NULL,
    s_address     VARCHAR(50) NOT NULL,
    s_city        VARCHAR(30) NOT NULL,
    s_nation      VARCHAR(30) NOT NULL,
    s_region      VARCHAR(30) NOT NULL,
    s_phone       VARCHAR(20) NOT NULL
) diststyle all;

CREATE TABLE customer 
(
    c_custkey     INTEGER     NOT NULL    sortkey,
    c_name        VARCHAR(50) NOT NULL,
    c_address     VARCHAR(30) NOT NULL,
    c_city        VARCHAR(30) NOT NULL,
    c_nation      VARCHAR(30) NOT NULL,
    c_region      VARCHAR(30) NOT NULL,
    c_phone       VARCHAR(30) NOT NULL,
    c_mktsegment  VARCHAR(30) NOT NULL
) diststyle all;

CREATE TABLE dwdate 
(
    d_datekey          INTEGER     NOT NULL   sortkey,
    d_date             VARCHAR(20) NOT NULL,
    d_dayofweek        VARCHAR(20) NOT NULL,
    d_month            VARCHAR(20) NOT NULL,
    d_year             INTEGER     NOT NULL,
    d_yearmonthnum     INTEGER     NOT NULL,
    d_yearmonth        VARCHAR(10) NOT NULL,
    d_daynuminweek     INTEGER     NOT NULL,
    d_daynuminmonth    INTEGER     NOT NULL,
    d_daynuminyear     INTEGER     NOT NULL,
    d_monthnuminyear   INTEGER     NOT NULL,
    d_weeknuminyear    INTEGER     NOT NULL,
    d_sellingseason    VARCHAR(20) NOT NULL,
    d_lastdayinweekfl  VARCHAR(1)  NOT NULL,
    d_lastdayinmonthfl VARCHAR(1)  NOT NULL,
    d_holidayfl        VARCHAR(1)  NOT NULL,
    d_weekdayfl        VARCHAR(1)  NOT NULL
) diststyle all;

CREATE TABLE lineorder 
(
    lo_orderkey        INTEGER     NOT NULL,
    lo_linenumber      INTEGER     NOT NULL,
    lo_custkey         INTEGER     NOT NULL,
    lo_partkey         INTEGER     NOT NULL   distkey,
    lo_suppkey         INTEGER     NOT NULL,
    lo_orderdate       INTEGER     NOT NULL   sortkey,
    lo_orderpriority   VARCHAR(20) NOT NULL,
    lo_shippriority    VARCHAR(20) NOT NULL,
    lo_quantity        INTEGER     NOT NULL,
    lo_extendedprice   INTEGER     NOT NULL,
    lo_ordertotalprice INTEGER     NOT NULL,
    lo_discount        INTEGER     NOT NULL,
    lo_revenue         INTEGER     NOT NULL,
    lo_supplycost      INTEGER     NOT NULL,
    lo_tax             INTEGER     NOT NULL,
    lo_commitdate        VARCHAR(20) NOT NULL,
    lo_shipmode        VARCHAR(20) NOT NULL
);

 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

In [93]:
# copying data into table
# %sql is to run a standalone sql command
# %%sql is to run a whole block as a sql block no more python
def loadTables(schema, tables):
    loadTimes = []
    SQL_SET_SCEMA = "SET search_path TO {};".format(schema)
    %sql $SQL_SET_SCEMA
    
    for table in tables:
        SQL_COPY  = """
                    copy {} from 's3://awssampledbuswest2/ssbgz/{}' 
                    credentials 'aws_iam_role={}'
                    gzip region 'us-west-2';
                    """.format(table, table, DWH_ROLE_ARN)

        print("======= LOADING TABLE: ** {} ** IN SCHEMA ==> {} =======".format(table, schema))
        print(SQL_COPY)

        t0 = time()
        %sql $SQL_COPY
        loadTime = time() - t0
        loadTimes.append(loadTime)

        print("=== DONE IN: {0:.2f} sec\n".format(loadTime))
    return pd.DataFrame({"table":tables, "loadtime_"+schema:loadTimes}).set_index('table')

In [94]:
# List of the tables to be loaded
tables = ["customer", "dwdate", "supplier", "part", "lineorder"]

In [101]:
# Insertion twice for each schema (WARNING!! EACH CAN TAKE MORE THAN 10 MINUTES!!!)
nodistStats = loadTables("nodist", tables)

 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
======= LOADING TABLE: ** customer ** IN SCHEMA ==> nodist =======

                   copy customer from 's3://awssampledbuswest2/ssbgz/customer' 
                   credentials 'aws_iam_role=arn:aws:iam::764499268961:role/dwh-role'
                   gzip region 'us-west-2';
                   
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
=== DONE IN: 13.98 sec

======= LOADING TABLE: ** dwdate ** IN SCHEMA ==> nodist =======

                   copy dwdate from 's3://awssampledbuswest2/ssbgz/dwdate' 
                   credentials 'aws_iam_role=arn:aws:iam::764499268961:role/dwh-role'
                   gzip region 'us-west-2';
                   
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
=== DONE IN: 1.45 sec

======= LOADING TABLE: ** supplier ** IN SCHEMA ==> nodist =

In [102]:
# Insertion twice for each schema (WARNING!! EACH CAN TAKE MORE THAN 10 MINUTES!!!)
distStats = loadTables("dist", tables)

 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
======= LOADING TABLE: ** customer ** IN SCHEMA ==> dist =======

                   copy customer from 's3://awssampledbuswest2/ssbgz/customer' 
                   credentials 'aws_iam_role=arn:aws:iam::764499268961:role/dwh-role'
                   gzip region 'us-west-2';
                   
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
=== DONE IN: 25.48 sec

======= LOADING TABLE: ** dwdate ** IN SCHEMA ==> dist =======

                   copy dwdate from 's3://awssampledbuswest2/ssbgz/dwdate' 
                   credentials 'aws_iam_role=arn:aws:iam::764499268961:role/dwh-role'
                   gzip region 'us-west-2';
                   
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
=== DONE IN: 1.97 sec

======= LOADING TABLE: ** supplier ** IN SCHEMA ==> dist =======

#### Loading Performance

In [ ]:
# Plotting of the timing results
stats = distStats.join(nodistStats)
#stats.plot.bar()
#plt.show()

In [113]:
stats.head(10)

loadtime_dist  loadtime_nodist
table                                    
customer   25.479078      13.978294      
dwdate     1.971943       1.448280       
supplier   15.436045      7.560319       
part       18.139198      5.215743       
lineorder  741.379035     628.864526

#### Query Performance

In [109]:
oneDim_SQL = """
             set enable_result_cache_for_session to off;
             SET search_path TO {};

             select sum(lo_extendedprice*lo_discount) as revenue
             from lineorder, dwdate
             where lo_orderdate = d_datekey
             and d_year = 1997 
             and lo_discount between 1 and 3 
             and lo_quantity < 24;
             """

twoDim_SQL = """
             set enable_result_cache_for_session to off;
             SET search_path TO {};

             select sum(lo_revenue), d_year, p_brand1
             from lineorder, dwdate, part, supplier
             where lo_orderdate = d_datekey
             and lo_partkey = p_partkey
             and lo_suppkey = s_suppkey
             and p_category = 'MFGR#12'
             and s_region = 'AMERICA'
             group by d_year, p_brand1
             """

drill_SQL = """
            set enable_result_cache_for_session to off;
            SET search_path TO {};

            select c_city, s_city, d_year, sum(lo_revenue) as revenue 
            from customer, lineorder, supplier, dwdate
            where lo_custkey = c_custkey
            and lo_suppkey = s_suppkey
            and lo_orderdate = d_datekey
            and (c_city='UNITED KI1' or
            c_city='UNITED KI5')
            and (s_city='UNITED KI1' or
            s_city='UNITED KI5')
            and d_yearmonth = 'Dec1997'
            group by c_city, s_city, d_year
            order by d_year asc, revenue desc;
            """


oneDimSameDist_SQL = """
                     set enable_result_cache_for_session to off;
                     SET search_path TO {};

                     select lo_orderdate, sum(lo_extendedprice*lo_discount) as revenue  
                     from lineorder, part
                     where lo_partkey  = p_partkey
                     group by lo_orderdate
                     order by lo_orderdate
                     """

def compareQueryTimes(schema):
    queryTimes  =[] 
    for i, query in enumerate([oneDim_SQL, twoDim_SQL, drill_SQL, oneDimSameDist_SQL]):
        t0 = time()
        q = query.format(schema)
        %sql $q
        queryTime = time() - t0
        queryTimes.append(queryTime)
    return pd.DataFrame({"query":["oneDim","twoDim", "drill", "oneDimSameDist"], "queryTime_"+schema:queryTimes}).set_index('query')

In [110]:
noDistQueryTimes = compareQueryTimes("nodist")
distQueryTimes = compareQueryTimes("dist") 
queryTimeDF = noDistQueryTimes.join(distQueryTimes)
#queryTimeDF.plot.bar()
#plt.show()

 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.
1 rows affected.
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.
280 rows affected.
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.
4 rows affected.
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.
2406 rows affected.
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.
1 rows affected.
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.
280 rows affected.
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.
4 rows affected.
 * postgresql://dwhuser:***@dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com:5439/dwh
Done.
Done.
2406 rows affected

queryTime_nodist  queryTime_dist
query                                           
oneDim          13.521617         7.288541      
twoDim          19.933780         15.266812     
drill           16.512507         15.760254     
oneDimSameDist  20.045202         13.208860

In [112]:
queryTimeDF.head(10)

queryTime_nodist  queryTime_dist
query                                           
oneDim          13.521617         7.288541      
twoDim          19.933780         15.266812     
drill           16.512507         15.760254     
oneDimSameDist  20.045202         13.208860

In [114]:
improvementDF = queryTimeDF["distImprovement"] = 100.0*(queryTimeDF['queryTime_nodist']-queryTimeDF['queryTime_dist'])/queryTimeDF['queryTime_nodist']
#improvementDF.plot.bar(title="% dist Improvement by query")
#plt.show()
improvementDF

query
oneDim            46.097121
twoDim            23.412359
drill             4.555655 
oneDimSameDist    34.104629
dtype: float64

### Cleanup

In [115]:
#### CAREFUL!!
#-- Uncomment & run to delete the created resources
redshift.delete_cluster( ClusterIdentifier=DWH_CLUSTER_IDENTIFIER,  SkipFinalClusterSnapshot=True)
#### CAREFUL!!

{'Cluster': {'ClusterIdentifier': 'dwh-cluster',
  'NodeType': 'dc2.large',
  'ClusterStatus': 'deleting',
  'MasterUsername': 'dwhuser',
  'DBName': 'dwh',
  'Endpoint': {'Address': 'dwh-cluster.cgjrwscs7tjx.us-west-2.redshift.amazonaws.com',
   'Port': 5439},
  'ClusterCreateTime': datetime.datetime(2019, 6, 4, 18, 39, 40, 858000, tzinfo=tzutc()),
  'AutomatedSnapshotRetentionPeriod': 1,
  'ManualSnapshotRetentionPeriod': -1,
  'ClusterSecurityGroups': [],
  'VpcSecurityGroups': [],
  'ClusterParameterGroups': [{'ParameterGroupName': 'default.redshift-1.0',
    'ParameterApplyStatus': 'in-sync'}],
  'ClusterSubnetGroupName': 'default',
  'VpcId': 'vpc-ce44eeb6',
  'AvailabilityZone': 'us-west-2d',
  'PreferredMaintenanceWindow': 'sun:09:00-sun:09:30',
  'PendingModifiedValues': {},
  'ClusterVersion': '1.0',
  'AllowVersionUpgrade': True,
  'NumberOfNodes': 4,
  'PubliclyAccessible': True,
  'Encrypted': False,
  'Tags': [],
  'EnhancedVpcRouting': False,
  'IamRoles': [{'IamRoleArn'

In [ ]:
myClusterProps = redshift.describe_clusters(ClusterIdentifier=DWH_CLUSTER_IDENTIFIER)['Clusters'][0]
prettyRedshiftProps(myClusterProps)

In [121]:
#### CAREFUL!!
#-- Uncomment & run to delete the created resources
iam.detach_role_policy(RoleName=DWH_IAM_ROLE_NAME, PolicyArn="arn:aws:iam::aws:policy/AmazonS3ReadOnlyAccess")
iam.delete_role(RoleName=DWH_IAM_ROLE_NAME)
#### CAREFUL!!

{'ResponseMetadata': {'RequestId': '04b53606-5937-4b8a-a6ea-85ce6778724e',
  'HTTPStatusCode': 200,
  'HTTPHeaders': {'x-amzn-requestid': '04b53606-5937-4b8a-a6ea-85ce6778724e, 04b53606-5937-4b8a-a6ea-85ce6778724e',
   'date': 'Tue, 04 Jun 2019 19:57:10 GMT',
   'content-type': 'text/xml',
   'content-length': '200'},
  'RetryAttempts': 0}}